source/libuv/uv_event_loop.c

```diff
@@ -145,33 +167,94 @@ static uint64_t s_timestamp_to_uv_millis(struct aws_event_loop *event_loop, cons
     return time_to_run;
 }
 
-static void s_uv_poll_cb(uv_poll_t *handle, int status, int events) {
+static void s_uv_poll_cb(uv_loop_t *loop, uv__io_t *w, unsigned int events) {
+
+    (void)loop;
 
+    uv_poll_t *handle = AWS_CONTAINER_OF(w, uv_poll_t, io_watcher);
     struct handle_data *handle_data = handle->data;
     int aws_events = 0;
 
-    uv_os_fd_t fd = 0;
-    if (uv_fileno((uv_handle_t *)handle, &fd)) {
-        return;
-    }
-
-    if (status >= 0) {
-        if (events & UV_DISCONNECT) {
-            aws_events |= AWS_IO_EVENT_TYPE_CLOSED;
-        }
-        if (events & UV_READABLE) {
+    if (events & POLLERR) {
+        uv_poll_stop(handle);
+        aws_events = AWS_IO_EVENT_TYPE_ERROR;
+    } else {
+        if (events & POLLIN) {
             aws_events |= AWS_IO_EVENT_TYPE_READABLE;
         }
-        if (events & UV_WRITABLE) {
+        if (events & POLLOUT) {
             aws_events |= AWS_IO_EVENT_TYPE_WRITABLE;
         }
-    } else {
-        aws_events = AWS_IO_EVENT_TYPE_ERROR;
+        if (events & POLLRDHUP) {
+            aws_events |= AWS_IO_EVENT_TYPE_REMOTE_HANG_UP;
+        }
+        if (events & POLLHUP) {
+            aws_events |= AWS_IO_EVENT_TYPE_CLOSED;
+        }
     }
 
     handle_data->on_event(handle_data->event_loop, handle_data->owner, aws_events, handle_data->on_event_user_data);
 }
 
+static void s_do_subscribe(struct libuv_loop *impl, struct handle_data *handle_data) {
+
+    /* Initalize the handle to default state */
+    uv_poll_init(impl->uv_loop, &handle_data->poll, handle_data->owner->data.fd);
+    handle_data->poll.data = handle_data;
+
+    /* Overwrite watcher callback because the default one filters the events down to just readable & writable */
+    handle_data->poll.io_watcher.cb = s_uv_poll_cb;
+
+    /* Set internal flags to a) get edge triggering and b) get remote hangup notifications */
+    handle_data->poll.io_watcher.pevents |= EVENT_LOOP_FLAGS;
+
+#if UV_VERSION_MAJOR == 0
+    /* uv won't let us tell it we care about remote hangups, so we have to trick it into thinking it wants it */
+    handle_data->poll.io_watcher.pevents |= POLLRDHUP;
+#else
+    /* Can't use RDHUP trick here, since uv will turn it off in uv__io_stop */
+    handle_data->uv_events |= UV_DISCONNECT;
+#endif
+
+    /* Start listening for events */
+    uv_poll_start(&handle_data->poll, handle_data->uv_events, (uv_poll_cb)0xBAADF00D);
```

Comment:
Is this value special to libuv, or just obvious? Would be good to make it clearer.